remote/remote_test.go

```diff
@@ -501,6 +502,52 @@ func testImage(t *testing.T, when spec.G, it spec.S) {
 				})
 			})
 		})
+
+		when("retry logic", func() {
+			var mockServer *h.MockServer
+
+			when("manifest API in registry returns status code 200", func() {
+				it.Before(func() {
+					mockServer, repoName = h.SetUpMockServer(t, "org/do-not-retry", http.StatusOK, 0)
+				})
+
+				it.After(func() {
+					mockServer.Server().Close()
+				})
+
+				it("does not retry after status code 200", func() {
+					assertExpectedTries(t, mockServer, repoName, 1)
+				})
+			})
+
+			when("manifest API in registry returns status code 404", func() {
+				it.Before(func() {
+					mockServer, repoName = h.SetUpMockServer(t, "org/retry-not-found", http.StatusNotFound, 2)
+				})
+
+				it.After(func() {
+					mockServer.Server().Close()
+				})
+
+				it("retries after status code 404", func() {
+					assertExpectedTries(t, mockServer, repoName, 3)
+				})
+			})
+
+			when("manifest API in registry returns status code 401", func() {
```

Comment:
It would be nice to have a test where we `fail` `succeed` to show that it works and assert we call twice but actually succeed.

Reply:
Wait, @jabrown85  the current tests is actually doing that. 

`mockServer, repoName = h.SetUpMockServer(t, "org/retry-unauthorized", http.StatusUnauthorized, 2)` the mock server is initialized to failed 2 times, on the third invocation it will respond with a good response.

Then

`assertExpectedTries(t, mockServer, repoName, 3)` is verifying the remote path was called 3 times and inside that verification, I am checking the method was executed succeded

```
func assertExpectedTries(t *testing.T, mockServer *h.MockServer, repoName string, expectedCount int) {
	_, err := remote.NewImage(repoName, authn.DefaultKeychain, remote.WithPreviousImage(repoName))
	h.AssertNil(t, err). // Method was succeded 
	h.AssertEq(t, mockServer.ActualCount(), expectedCount)
}
```

Maybe I can save how many times the bad response was returned and assert against that value? also, that number can be calculated based on the failed times configuration and the total number of invocations too

Reply:
ah, ok. I didn't read that the next time would be successful. I think you are good to go